frame/bonded-finance/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+		///
+		/// Emits a `OfferCancelled`.
+		#[pallet::weight(10_000)]
+		pub fn cancel_offer(origin: OriginFor<T>, offer: T::BondOfferId) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			Self::do_cancel_offer(offer)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn account_id(offer: T::BondOfferId) -> AccountIdOf<T> {
+			T::PalletId::get().into_sub_account(offer)
+		}
+
+		pub fn do_cancel_offer(offer_id: T::BondOfferId) -> DispatchResult {
+			let (creator, _) =
+				BondOffers::<T>::try_get(offer_id).map_err(|_| Error::<T>::BondOfferNotFound)?;
+			let offer_account = Self::account_id(offer_id);
+			T::NativeCurrency::transfer(&offer_account, &creator, T::Stake::get(), true)?;
+			BondOffers::<T>::remove(offer_id);
+			Self::deposit_event(Event::<T>::OfferCancelled { offer: offer_id });
+			Ok(())
+		}
+
+		#[transactional]
+		pub fn do_offer(
+			from: &AccountIdOf<T>,
+			offer: BondOfferOf<T>,
+		) -> Result<T::BondOfferId, DispatchError> {
+			ensure!(
+				offer.valid(
+					<T::Vesting as VestedTransfer>::MinVestedTransfer::get(),
+					T::MinReward::get()
+				),
+				Error::<T>::InvalidBondOffer
+			);
+			let offer_id = BondOfferCount::<T>::mutate(|offer_id| {
+				*offer_id = offer_id.next();
+				*offer_id
+			});
+			let offer_account = Self::account_id(offer_id);
+			T::NativeCurrency::transfer(from, &offer_account, T::Stake::get(), true)?;
+			T::Currency::transfer(
+				offer.reward_asset,
+				from,
+				&offer_account,
+				offer.reward_amount,
+				true,
+			)?;
+			BondOffers::<T>::insert(offer_id, (from.clone(), offer));
+			Self::deposit_event(Event::<T>::NewOffer { offer: offer_id });
+			Ok(offer_id)
+		}
+
+		#[transactional]
+		pub fn do_bond(
+			offer_id: T::BondOfferId,
+			from: &AccountIdOf<T>,
+			contracts: BalanceOf<T>,
+		) -> Result<BalanceOf<T>, DispatchError> {
+			BondOffers::<T>::try_mutate(offer_id, |offer| {
+				offer
+					.as_mut()
+					.map(|(creator, offer)| {
+						ensure!(
+							offer.contracts > BalanceOf::<T>::zero(),
+							Error::<T>::OfferCompleted
+						);
+						ensure!(
+							contracts > BalanceOf::<T>::zero() && contracts <= offer.contracts,
+							Error::<T>::InvalidNumberOfContracts
+						);
+						// NOTE(hussein-aitlahcen): can't overflow, subsumed by `offer.valid()` in
+						// `do_offer`
+						let value = contracts * offer.price;
+						ensure!(
+							T::Currency::can_withdraw(offer.asset, from, value)
+								.into_result()
+								.is_ok(),
+							Error::<T>::NotEnoughAsset
+						);
+						let offer_account = Self::account_id(offer_id);
+						T::Currency::transfer(offer.asset, from, &offer_account, value, true)?;
+						let reward_share = T::Convert::convert(
+							multiply_by_rational(
+								T::Convert::convert(value),
+								T::Convert::convert(offer.reward_amount),
+								// NOTE(hussein-aitlahcen): checked by `offer.valid()` in
+								// `do_offer`
+								T::Convert::convert(offer.total_price().expect("impossible; qed;")),
```

Comment:
contracts/total_contracts , you do not need to * price to times.

Reply:
Yep, this in an artifact, good catch :)